vsphere/resource_vsphere_license_test.go

```diff
@@ -0,0 +1,225 @@
+package vsphere
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"testing"
+
+	"regexp"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/license"
+)
+
+var (
+	testAccLabels = map[string]string{
+		"VpxClientLicenseLabel": "Hello World",
+		"TestTitle":             "FooBar",
+	}
+)
+
+func TestAccVSphereLicenseBasic(t *testing.T) {
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			testAccVSpherePreLicenseBasicCheck(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testAccVSphereLicenseDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccVSphereLicenseBasicConfig(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccVSphereLicenseExists("vsphere_license.foo"),
+				),
+			},
+		},
+	})
+
+}
+
+func TestAccVSphereLicenseInvalid(t *testing.T) {
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+		},
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccVSphereLicenseInvalidConfig(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccVSphereLicenseNotExists("vsphere_license.foo"),
+				),
+				ExpectError: regexp.MustCompile("License file not found"),
+			},
+		},
+	})
+
+}
+
+func TestAccVSphereLicenseWithLabels(t *testing.T) {
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			testAccVSpherePreLicenseBasicCheck(t)
+		},
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccVSphereLicenseWithLabelConfig(testAccLabels),
+				Check: resource.ComposeTestCheckFunc(
+					testAccVSphereLicenseWithLabelExists("vsphere_license.foo"),
+				),
+			},
+		},
+	})
+
+}
+
+func testAccVSphereLicenseInvalidConfig() string {
+
+	// quite sure this key cannot be valid
+	return `resource "vsphere_license" "foo" {
+  					license_key = "HN422-47193-58V7M-03086-0JAN2"
+			}`
+}
+
+func testAccVSphereLicenseWithLabelConfig(labels map[string]string) string {
```

Comment:
We are getting close here - remember though that we don't necessarily need to render the labels here dynamically - it's better to just use a static string, and roll the function into a much more simpler form:

```go
func testAccVSphereLicenseWithLabelConfig() string {
  return fmt.Sprintf(`
resource "vsphere_license" "foo" {
  license_key = "%s"

  labels {
    VpxClientLicenseLabel = "Hello World"
    TestTitle = "FooBar"
  }  
}
`, os.Getenv("VSPHERE_LICENSE"))
}
```